### ISSUE TITLE
Support General Broadcasting

### ISSUE THREAD

**Issue by agibsonccc** [deeplearning4j/libnd4j#513]
Recently in nd4j, I implemented the ability to do automatic broadcasting. I figured it was time to apply it to libnd4j as well:
https://docs.scipy.org/doc/numpy-1.10.1/user/basics.broadcasting.html

Basic changes needed:
The ability to create a new output array with a new shape (right now we only broadcast along x)

A general broadcast would allow looping over both arrays and outputting new elements till the new array is filled.

Basically what we would need is filling in a TAD along the new result array instead of x.

**Comment by raver119** [comment]
Can you please point us to your new java impl?

**Comment by agibsonccc** [comment]
https://github.com/deeplearning4j/nd4j/blob/9a3867980f49f149866e73800c98e5d4a242511a/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ndarray/BaseNDArray.java#L3293

using:
Valid broadcast dimensions:
https://github.com/deeplearning4j/nd4j/blob/36f7e5537fa79b0ada81fd2239b0a4baf789a8e8/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/shape/Shape.java#L73-L72

Output shape:
https://github.com/deeplearning4j/nd4j/blob/36f7e5537fa79b0ada81fd2239b0a4baf789a8e8/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/shape/Shape.java#L110

**Comment by raver119** [comment]
Implemented
